### PR TITLE
IDEMPIERE-5443 - Date Range Filter Component

### PR DESCRIPTION
**Pull request by PeterTakacs300**
JIRA ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-5443

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [ ] I have made corresponding changes to the documentation as follows:
- - [ ] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

**Comment by PeterTakacs300** [comment]
Current state:

https://user-images.githubusercontent.com/93127072/195288289-b8c88d1e-07bf-4b18-b5ac-d5fe2d61a62c.mp4


We would like to implement a new editor type, where we can store two values, and display it as one value (it would be different from the current solution, where if IsRange = 'Y' then it displays two separate editors with a single value) - similar to this:

https://user-images.githubusercontent.com/93127072/195289860-f056523a-3e50-4456-a7e0-225f2dc6852c.mp4

**Comment by igorpojzl** [comment]
Hello,
To Implement Range Editor omponent:
We Looked at WEditor class if we can expand it, so it can be used as Range Component, but I think it is too complicated to implement there. I think Gridfield class should not be changed as its holding one value.
Possible Solution is to Create new Abstract class for Editor Range Component, with its Factory and implement its Rendering to  ProcessParameterPanel, FindWindow and Info Window.
We will store 2 Gridfield objects in this class. It will store 2 Values and One Display Value.

**Comment by hengsin** [comment]
hi @igorpojzl , I've not exploit this much but looking at your current implementation, I imaging it could be a component that manage 2 WEditor component (and it doesn't even has to be a WEditor component itself).

**Comment by igorpojzl** [comment]
@hengsin So do you suggest to create "Middle" Component containing 2 WEditors. But we dont want to render either of those WEditors and then connect events and ProcessParameterPanel to those 2 WEditors + GridFields? And Use WEditors only to store final data?

**Comment by norbertbede** [comment]
ok. for me is important to approve design, before continue avoid double work. 
so if this way is suggested/acceptable for commit (after review), then we can implement it in next days.

File: migration/iD10/postgresql/202210041053_IDEMPIERE-5443.sql [hunk 1]
```diff
@@ -0,0 +1,175 @@
+-- CLDE-748
+SELECT register_migration_script('202210041053_PlaceholderForTicket.sql') FROM dual;
```
Comment (CarlosRuiz-globalqss) [inline migration/iD10/postgresql/202210041053_IDEMPIERE-5443.sql:2]: filename is wrong here

**Review by CarlosRuiz-globalqss — CHANGES_REQUESTED** [review]

File: migration/iD10/postgresql/202210041053_IDEMPIERE-5443.sql (hunk 1 above)
Comment (PeterTakacs300) [reply migration/iD10/postgresql/202210041053_IDEMPIERE-5443.sql]: fixed

**Review by CarlosRuiz-globalqss — APPROVED** [review]

**Review by CarlosRuiz-globalqss — CHANGES_REQUESTED** [review]
my tests went OK - changing back to request changes as this is still in design discussion in the ticket and is mentioned to be still incomplete as a POC.

**Comment by hengsin** [comment]
This latest change confuse me -why can't we use the find window approach to process parameter ?

**Comment by PeterTakacs300** [comment]
Date Range Picker editor (when reference Date Range Picker is selected):
https://user-images.githubusercontent.com/93127072/196622276-e2384ce0-9f0c-4073-84b2-4b313788663d.mp4
- implemented display logic, mandatory logic, read only logic
- the popup shows only after the calendar icon is clicked, the text field itself is read only, shows the selected date range as human-readable string

Date Range Picker Button (shows when reference type is date type and IsRange='Y'):
https://user-images.githubusercontent.com/93127072/196622283-151b3699-e996-43d2-87f6-7cc86089f601.mp4
- implemented read only logic (button disabled)

**Comment by norbertbede** [comment]
I believed, @PeterTakacs300 and @igorpojzl understood the previous communication design. Please be more specific, what approach you suggest. exact example or existing comment (oversight.)

**Comment by hengsin** [comment]
hi @PeterTakacs300 , that's clearer now but why do we need 2 implementation for essentially the same thing here ?

**Comment by PeterTakacs300** [comment]
Hi @hengsin, 

These are basically two use cases of the new DateRangePicker popup-component.

1. add it as an improvement for all the Date process parameters where IsRange = 'Y' (through DateRangeButton)
2. create a new special "editor" (DateRangeEditor + register a new AD_Reference: Date Range Picker) - so we will be able to display date ranges in a more user friendly way - it creates also possibilities for future development/customisation and modernisation of this editor

**Comment by hengsin** [comment]
Fine, perhaps it is good to have more UI option. Anyway, do note that it is best not to lump > 1 use case/enhancement into 1 ticket since that will complicate review and testing.

**Comment by PeterTakacs300** [comment]
We should define then, what tickets should we create, and what will be their content

1. ticket would contain only the improvement of the current IsRange='Y' Date process parameters
2. ticket would have the new AD_Reference and the new DateRangeEditor for the process parameters - it's prerequisite would be the merging of the first ticket

So should I first remove the DateRangeEditor and related code from this PR, and move it to a separate one with a new ticket, after this PR is merged?

**Comment by hengsin** [comment]
Hi @PeterTakacs300 ,

Yes, I think it is better to move 2 to a separate ticket. That potentially needs more review and discussion. The enhancement for find window and isRange process parameter looks ready for merge.

**Comment by PeterTakacs300** [comment]
Hi @hengsin , I've removed the editor from this PR as we agreed and I will commit it to a new one with a new ticket after this PR is merged

**Comment by hengsin** [comment]
hi @PeterTakacs300, any plan to implement this for the find window as well ?

**Comment by PeterTakacs300** [comment]
We decided for now to implement it only for process parameters

**Comment by hengsin** [comment]
[pr1526.patch.txt](https://github.com/idempiere/idempiere/files/9846136/pr1526.patch.txt)
- minor refactoring
- add implementation for find window

**Comment by PeterTakacs300** [comment]
Hi @hengsin, patch applied

**Comment by norbertbede** [comment]
hi @CarlosRuiz-globalqss  
we review more bugs, @PeterTakacs300 will provide patch with changes (hope today or tomorrow morning)
thanks for testing.
n

**Comment by PeterTakacs300** [comment]
fixes and changes:

- fixed current week filling wrong dates
- fixed Between to not allow selecting an end date before the start date
- refactored Previous and Next modes:
- - after refactor they return calendar ranges (similar to Current)
- - e.g. before the refactor Last Month returned the last 31 days, now it returns the last whole month (e.g. 09/01/2022 - 09/30/2022) not including this month by default
- refactored the generic _getInterval_ method
- UI changes:
- - removed Clear button
- - renamed Update Filter button to Apply + added some style
- - the Selected Time Range TextBox is moved next to the Apply button, the label is renamed simply to Range to save some space
- - when popup first opened, it is displayed "after center" instead of "after end" and we decided to set the picker to Current Month by default, instead of the Quick mode
- - fixed css bug on Between mode (when changing months with the arrows on one calendar, the other calendar was jumping unexpectedly)
- - the Spinner component on Previous and Next modes is changing the selected range dynamically (we don't have to click outside the component to apply the changes)

**Comment by CarlosRuiz-globalqss** [comment]
Thanks @PeterTakacs300 

> fixed current week filling wrong dates

Still not working, today is showing:
10/24/2022 - 10/23/2022
and it shows error when Applying

> fixed Between to not allow selecting an end date before the start date

Works perfect


New issue:

Tried "Before" and got a NPE:
```
java.lang.NullPointerException
	at java.base/java.util.Date.getMillisOf(Date.java:956)
	at java.base/java.util.Date.before(Date.java:915)
	at org.adempiere.webui.window.DateRangePicker.lambda$0(DateRangePicker.java:156)
```

**Comment by PeterTakacs300** [comment]
Hi @CarlosRuiz-globalqss, both issues should work now - first day of week is set based on the locale, it defines weather we start the current week on Sunday or Monday

**Review by CarlosRuiz-globalqss — APPROVED** [review]
tests are passing OK in test-feature server

**Pull request merged**